MekHQ/src/mekhq/campaign/Campaign.java

```diff
@@ -1516,30 +1520,73 @@ public void addPartWithoutId(Part p) {
                         ((AmmoStorage) spare).changeShots(((AmmoStorage) p)
                                 .getShots());
-                        return;
+                        mergedWith = spare;
                     }
                 } else {
                     spare.incrementQuantity();
-                    return;
+                    mergedWith = spare;
                 }
             }
         }
 
-        parts.put(p.getId(), p);
-        MekHQ.triggerEvent(new PartNewEvent(p));
+        // If we weren't merged we are being added
+        if (null == mergedWith) {
+            parts.put(p.getId(), p);
+            MekHQ.triggerEvent(new PartNewEvent(p));
+        } else {
+            // Go through each unit and its refits to see if the new armor ID should be updated
+            // CAW: I believe all other parts on a refit have a unit assigned to them.
+            for (Unit u : getUnits()) {
+                Refit r = u.getRefit();
+                // If there is a refit and this part matches the new armor, update the ID
+                if (null != r) {
+                    if (mergedWith instanceof Armor
+                        && r.getNewArmorSuppliesId() == p.getId()) {
+                        MekHQ.getLogger().info(Campaign.class, "addPartWithoutId",
+                            String.format("%s (%d) was merged with %s (%d) used in a refit for %s", p.getName(),
+                                p.getId(), mergedWith.getName(), mergedWith.getId(), u.getName()));
+                        Armor mergedArmor = (Armor)mergedWith;
+                        r.setNewArmorSupplies(mergedArmor);
+                    } else {
+                        List<Integer> ids = r.getOldUnitPartIds();
+                        for (int ii = 0; ii < ids.size(); ++ii) {
+                            int oid = (int)ids.get(ii);
+                            if (p.getId() == oid) {
+                                MekHQ.getLogger().info(Campaign.class, "addPartWithoutId",
+                                String.format("%s (%d) was merged with %s (%d) used in the old unit in a refit for %s", p.getName(),
+                                    p.getId(), mergedWith.getName(), mergedWith.getId(), u.getName()));
+                                ids.set(ii, mergedWith.getId());
+                            }
+                        }
+                        ids = r.getNewUnitPartIds();
+                        for (int ii = 0; ii < ids.size(); ++ii) {
+                            int nid = (int)ids.get(ii);
+                            if (p.getId() == nid) {
+                                MekHQ.getLogger().info(Campaign.class, "addPartWithoutId",
+                                String.format("%s (%d) was merged with %s (%d) used in the new unit in a refit for %s", p.getName(),
+                                    p.getId(), mergedWith.getName(), mergedWith.getId(), u.getName()));
+                                ids.set(ii, mergedWith.getId());
```

Comment:
What's the point of adding elements to the ids collection if you're not using it? Is it just a debugging thing?

Reply:
Those are loaded at CPNX read and eventually get out of sync. I'm not sure if these IDs getting out of sync are the proximate cause, but it seemed worth doing if we can.